### ISSUE TITLE
Spec split Tokenizer and TreeConstructor

### ISSUE THREAD

**Issue by jacobrheath** [heathj/gobrowse#16]
I did my best to split these into completely separate components so they could run in parallel, but the spec tighly couples them in a few places that makes this difficult. I am having to refactor these components so that the shared states are bridged in a few ways. Also I had to remove the parallel nature. I think I could make it work, but at the cost of a bunch of weird and exceptional complexity and probably not that much performance gain.

For example:

- In a handfull of cases the tree constructor needs to change the state of the tokenizer
- The tokenizer sometimes will need to know the current adjusted node when parsing [CDATA stuff
- Lots of other crossover

**Comment by heathj** [comment]
https://github.com/heathj/gobrowse/commit/f10c806b442947b0f81d57611f92bddeec933be7
